### DISCUSSION TITLE
Enabling buttons dynamically

### DISCUSSION THREAD

**Discussion by Roopesh-J** [discord-py-ui/discord-ui#103] [Q&A] [answered]
# Can buttons be enabled dynamically? 

I want to send a message with both a SelectMenu and Button component. However I want to have the button be disabled until something in the SelectMenu is chosen. Once an option is selected, the button can be updated to be enabled without having to resend the message or edit it. Is this possible?

**Comment by kvsxxx** [comment, marked as answer]
Well, yes but no. You can enable the button dynamically, but this will send a edit request to the api, so you would kind of edit the message. 

In the current version, this can be done by enabling **all** components in the message

```py
...
# send the message
msg = await ctx.send("this is a test", components=[
    # disable the button with disabled=True
    Button("hmm", "test", disabled=True), 
    SelectMenu("test", [SelectOption("2", "test")])
])
# wait for a selection
select = await msg.wait_for("select", bot)
# respond to the selected menu
await select.respond()
# enable all components, setting the disable parameter to False, 
# so for every component the `.disabled` property will be set to False, which means that they will be eanabled
await msg.disable_components(False)
```

If you want to make sure only the first component (the button) is going to be enabled, you would need to create a workaround for that, because in the current version enabling/disabling components is limited to setting all components


```py
msg = await ctx.send("this is a test", components=[
    # disable the button with disabled=True
    Button("hmm", "test", disabled=True), 
    SelectMenu("test", [SelectOption("2", "test")])
])
# wait for a selection
select = await msg.wait_for("select", bot)
# respond to the selected menu
await select.respond()

# make a "backup" for the components
components = msg.components
# enable the first component
components[0].disabled = False
# edit the message and update the components
await msg.edit(components=components)
```

- - -

If you're using the [__preview version__](https://github.com/discord-py-ui/discord-ui/tree/preview) you don't have to use the workaround for just enabling the first component

```py
# send the message
msg = await ctx.send("this is a test", components=[
    # disable the button with disabled=True
    Button("hmm", "test", disabled=True), 
    SelectMenu("test", [SelectOption("2", "test")])
])
# wait for a selection
select = await msg.wait_for("select", bot)
# respond to the selected menu
await sel.send()
# enable the first component
await msg.disable_component(0, False)
```

**Reply by Roopesh-J** [reply]
That first code block is perfect for my situation. I knew I had to edit the message but before I would just resend all the components but that resets the user's choice of the dropbox. '.disable_component()' is exactly what I am looking for I think. Thank you!

**Reply by Roopesh-J** [reply]
So turns out any edit to a message will reset the SelectMenu choice, atleast with everything I have tried. I was going to try just setting the default option of the Menu to be the users choice before the edit but the SelectMenu allows up to 4 choices and so some information could be lost. Should I just have the buttons be a separate message or just always enabled?

**Reply by kvsxxx** [reply]
Okay sorry I didn't see that there were more than one option that could be selected

You can mark more than 1 option as a default option

```py
   msg = await ctx.send("this is a test", components=[
        # disable the button with disabled=True
        Button("hmm", "test", disabled=True), 
        SelectMenu("test", [SelectOption("2", "test"), SelectOption("3", "test"), SelectOption("4", "test"), SelectOption("5", "test")], max_values=4)
    ])
    # wait for a selection
    select = await msg.wait_for("select", bot)
    # respond to the selected menu
    await select.respond()
    
    # make a "backup" for the components
    components = msg.components
    # enable the first component
    components[0].disabled = False
    
    # for every index of a selected option
    for index in [i for i, o in enumerate(components[1].options) if o.value in select.selected_values]:
        # mark option as a default option in the select menu
        components[1].set_default_option(index)
    
    # edit the message and update the components
    await msg.edit(components=components)
```

**Reply by Roopesh-J** [reply]
I tried something similar but it just keep resetting the default option instead of adding to it. However I tried doing it in the SelectMenu parameter itself so that could be why. This is perfect, thank you, you're so helpful!
